### ISSUE TITLE
What is the best way to return custom validation error messages

### ISSUE THREAD

**Issue by arvind-agarwal** [exegesis-js/exegesis#9]
If the validation is part of OAS spec than it gives a nice Error object e.g.,

```
{
   "message" : "Validation errors",
   "errors" : [
      {
         "location" : {
            "path" : [],
            "name" : "message",
            "docPath" : [
               "paths",
               "/hello",
               "get",
               "parameters",
               "0"
            ],
            "in" : "query"
         },
         "type" : "error",
         "message" : "Missing required query parameter \"message\""
      }
   ]
}
```

What is the best way to do additional validations (more complex ones) in controllers and return the error in similar format, obviously I can build it manually but ideally if the framework has a standard way to send such errors that would help in being consistent.

**Comment by arvind-agarwal** [comment]
@jwalton thanks for the super fast response.

context.makeError sounds perfect, I did try it but it didn't give me the error in the same format. So if it can be enhanced for validation errors it would be perfect.

Your comments about Path and type also seems very logical.

**Comment by jwalton** [comment]
Fixed in 1.0.0-rc0.  There's now a `context.parameterLocations`, an I added a `context.makeValidationError(message, location)`.

Other notably changes in 1.0.0-rc0:

* authenticators may no longer return `type: 'fail'`.  Instead they should return `type: 'missing'` if thre credentials are not present, or `type: 'invalid'` if the credentials are there but wrong (i.e. wrong password, missing user, etc...).  If we try to auth a request and we find a 'missing' we'll move on to the next authentication type, but if it's 'invalid' we'll stop trying other authenticators - the user tried to authenticate, but they're obviously not getting it right.

* `path` and `docPath` in validation errors are now JSON Pointers instead of arrays.

* Plugins are now passed the complete set of options that were passed into exegesis.  (Later I'm planning on adding a feature where you can specify a custom function to call to generate validation errors, in case you want to do something crazy like generate XML - would be handy if this was available to plugins.)

**Comment by jwalton** [comment]
(And I just added a 1.0.0-rc1, which renames `context.body` to `context.requestBody`, mainly for the sanity of Koa users.  :) )
